crates/bevy_ecs/src/system/system_param.rs

```diff
@@ -371,6 +378,13 @@ impl<'a, T: Resource> SystemParam for Res<'a, T> {
 // conflicts with any prior access, a panic will occur.
 unsafe impl<T: Resource> SystemParamState for ResState<T> {
     fn init(world: &mut World, system_meta: &mut SystemMeta) -> Self {
+        // If the resource can't be shared between threads, make sure it runs on the main thread.
+        // FIXME: We should be able to register access as 'exclusive reads' to prevent shared access
+        // without forcing the system to run on the main thread.
```

Comment:
IMO, there's no point to merging this PR as is without this. You would otherwise just use a `NonSend` for the same effect.

You can treat a `Res<T: !Sync>` the same as `ResMut<T: !Sync>`. Move the `if !T::IS_SYNC` down below, and change the non-Sync branch to add a write instead.

Reply:
> IMO, there's no point to merging this PR as is without this. You would otherwise just use a NonSend for the same effect.

I want to keep this PR as small as possible, so it's easier to review. If this gets merged, I'll open a follow-up PR that makes the `Access` changes.

These changes are easily separable, so I see no reason to roll them up into one PR. Especially since the `Access` changes are just an optimization.

Also, this already allows `ResMut<T: !Sync>` to be used off of the main thread.

> You can treat a Res<T: !Sync> the same as ResMut<T: !Sync>. Move the if !T::IS_SYNC down below, and change the non-Sync branch to add a write instead.

This won't work, it will mess up ambiguity detection.